#### NLTK: natural language ?? (library)
#### python은 2.7 ver-> 3.0
#### change가 많아서 호환이 잘 안됨
#### python 사용자 중 2.7에 머물기도 해
#### re: regular expression 관련 패키지
#### function 사용된 거: re.search, re.findall 이 정도만 사용한 상태
#### bracket 해서 list comprehension에 익숙해질 것
#### 코드창인데 markdown처럼 comment를 넣으려면 앞에 # 넣어도 ok
#### range로 #을 넣는 기능을 위해서(너무 계속 # 붙이려면 귀찮으니까)  '''~''' or """~""" (적어놓았으나 실행되진 않음)

In [4]:
nltk.download()

NameError: name 'nltk' is not defined

In [1]:
import nltk
import re

In [21]:
var="""
Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures
.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]      Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	        Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	    Parentheses that indicate the scope of the operators
(...)       Matches whatever regular expression is inside the parentheses
"""

#### 명심: string에 적용되는 함수이다
#### string에 적용되는 것이 regexp
#### 사전을 통째로  nltk에서 가져옴

In [26]:
dic = nltk.corpus.words.words('en')
print(type(dic))

<class 'list'>


In [14]:
a=[]
for w in nltk.corpus.words.words('en'):
    if w.islower():
        a.append(w)

#### in 뒤의 list
#### 그 리스트 안의 단어가 소문자일 때만 wordlist 안의 리스트에 쌓인다.
#### 최초 10개만 보여다오

In [27]:
wordlist = [w for w in dic if w.islower()]
print(wordlist[:10])

['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


In [33]:
wordlist = [w for w in nltk.corpus.words.words('en') if w.islower()]
print(wordlist[:10])

['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


In [31]:
a==wordlist

True

#### regexp -string에 적용된다, list에 적용되지 않는다
#### for loop의 의미: 어떤 사전이 있는데 이 조건에 해당하는 단어를 찾아라(X)
#### regexp를 리스트 전체에 적용시키기 위해서는 한 요소씩(하나하나씩) 적용해야 한다(하나하나는 string이니까)

#### 첫째: 찾는 조건, 둘째: 찾는 대상(리스트도 될 수 있고, string도 들어갈 수 있다)
#### 뒤어세 앞의 내용을 찾아라
#### 보여주는 방식: 찾은 걸 보여줄 수 있고, 찾은 range를 보여줄 수도 있다
#### 여기선 둘 다 나옴

In [19]:
re.search('abc','abcde')   # 함수의 입력 갯수: 2개
print(type(re.search('abc','abcde')))

<class 're.Match'>


#### output으로 아무것도 없는 이유는 2째 input에 1째 input에 해당하는 부분이 없으니까

In [18]:
re.search('ed','abcde')

In [17]:
x = re.search('abc','abcde')
print(x)

<re.Match object; span=(0, 3), match='abc'>


#### wordlist 안에 있는 단어들이 w로 들어가서 for loop을 돈다(횟수: len(wordlist))
#### w에 쌓일지 말지는 if문에서 결정됨 -ed로 끝나는 단어 가져오기, ed로 끝나지 않으면 그냥 패스하기
#### 그렇게 해서 만들어진 리스트 중 처음부터 10번째 것까지만 result에 들어간다.
#### 이럴 때 if 문이 없다면([:10]을 입력하지 않았다고 가정하자) result == wordlist

In [18]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result)
result = [w for w in wordlist if re.search('ed$', w)]
print(result[:10])

['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']
['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']


In [ ]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result[:10])  # 여기서 10은 redundant

#### ^: 앞, $: 뒤
#### 3번째 character가 j, 6번째 character가 t, 그리고 t 뒤에 2개의 character가 있는 단어를 찾아라(각각의 .== 각각 하나의 character)

In [21]:
result = [w for w in wordlist if re.search('^..j..t..$', w)]    # 출력되는 리스트의 단어릐 length: 8
print(result[:10])
print(result)
print(len(result))

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']
['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']
13


In [47]:
result = [w for w in wordlist if re.search('^..j..t..$', w)][:10]    # 출력되는 리스트의 단어릐 length: 8
print(result[:10])
print(type(result))

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']
<class 'list'>


In [48]:
result = [w for w in wordlist if re.search('^..j....t..$', w)][:10]
print(result[:10])

['adjacently', 'objurgator', 'unjacketed', 'unjesuited']


#### 첫째 자리: g,h,i 중 하나
#### 둘째 자리: m,n,o 중 하나
#### 셋째 자리: j,l,k 중 하나
#### 넷째(=마지막자리($이 입력되어 있으므로)) 자리: d,e,f 중 하나
#### 인 단어를 찾아서 앞의 10개까지만 result에 넣어라

In [7]:
result = [w for w in wordlist if re.search('^[ghi][mno][jlk][def]$', w)][:10]
print(result[:10])

['gold', 'golf', 'hold', 'hole']


#### 첫째 자리: a 또는 h
#### 둘째 자리~끝 자리: 아무것도 오지 않을 수 있고, 온다면 a또는 h가 온다.
#### +의 의미: 1회 이상
#### 10개까지 print하라고 했으나 원래 변수의 길이가 10보다 짧으면 짧은 대로 print된다.

In [8]:
result = [w for w in wordlist if re.search('^[ah]+$', w)][:10]
print(result[:10])

['a', 'aa', 'ah', 'aha', 'h', 'ha', 'hah']


#### 너무 많아서 끝 부분이 ...으로 나옴:  ('3.80', ...])
#### world street journal
#### wsj에서 나온 단어들이 나온 것
#### type(wsj) == list

In [4]:
wsj = sorted(set(nltk.corpus.treebank.words()))
# wsj

#### 처음(^이 있으므로): 0~9 중 하나(+: 1회 이상 올 수 있다)
### \ . :  특수문자인 점을 print하는 방법(실제 .을 찾는 방법)
#### 끝(달러모양): 0~9 중 하나(+: 1회 이상 올 수 있다)
#### 만약 [09] -> 0 또는 9
### \ + : 특수문자인 +을 찾아라
#### 실수를 찾는 셈(음수 제외)
#### '^'와 '$' 없애면 정해진 패턴이 아무데서나 포함되기만 하면 됨

In [25]:
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
print(wordlist[:10])
print(wordlist[470:])

['0.0085', '0.05', '0.1', '0.16', '0.2', '0.25', '0.28', '0.3', '0.4', '0.5']
['9.9', '92.9', '93.3', '93.9', '94.2', '94.8', '95.09', '96.4', '98.3', '99.1', '99.3']


#### 알파벳이 뭐가 나오든!
#### 알파벳이든 숫자든 자리 1개 = .
#### 왜 달러모양이 2개? '\달러모양'의 의미: 특수문자인 달러모양기호가 붙어있다
#### 알파벳 아무거나 + $로 끝나는 것을 찾아라
#### [A-Z0-9]: 알파벳 또는 숫자 1개 나옴.(특수문자를 제외한 것) len = 1
#### [^abc] : a,b,c가 아닌 것 아무거나 중 1개만 나와라

In [27]:
wordlist = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
print(wordlist[:10])

['C$', 'US$']


#### 리스트가 비어 있다:1개이상의 소문자 뒤에 달러모양이 붙은 경우는 없다

In [28]:
wordlist = [w for w in wsj if re.search('^[a-z]+\$$', w)]
print(wordlist[:10])

[]


#### 결과의 의미: 알파벳이 1개 이상 나오고 뒤에 달러모양이 오는 경우는 대문자 1~2개 뒤에 달러모양이 오는 경우만 있다.

In [29]:
wordlist = [w for w in wsj if re.search('^[A-Za-z]+\$$', w)]
print(wordlist[:10])

['C$', 'US$']


#### {숫자} : 반복 횟수
#### 0~9에서 하나의 숫자의 골라 나오는 과정이 4번 반복된다.

In [12]:
wordlist = [w for w in wsj if re.search('^[0-9]{4}$', w)]
print(wordlist[:10])

['1614', '1637', '1787', '1901', '1903', '1917', '1925', '1929', '1933', '1934']


#### 리스트 요소의 특징
#### 처음에 숫자가 하나 이상 나온다
#### -: 그냥 입력되고 출력된다(그냥 특수기호일 뿐이라 별다른 의미가 없다)
#### 끝엔 알파벳이 3~5자리가 나온다
#### 하나 이상이라고 지정했는데 하나짜리는 없나봐(숫자)

In [13]:
wordlist = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
print(wordlist[:10])

['10-day', '10-lap', '10-year', '100-share', '12-point', '12-year', '14-hour', '15-day', '150-point', '190-point']


#### <이 리스트 요소의 특징>
#### 1) 알파벳이 5개 이상
#### 2) -가 쓰인다
#### 3) 알파벳이 2~3개
#### 4) -가 쓰임
#### 5) 알파벳이 6개 이하인 것
#### 하나의 결과물을 위해서 다양한 regexp을 쓸 수 있다

In [14]:
wordlist = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]
print(wordlist[:10])

['black-and-white', 'bread-and-butter', 'father-in-law', 'machine-gun-toting', 'savings-and-loan']


#### | : '또는'
#### ed로 끝나거나 ing로 끝나는 것
#### (||||) 이렇게 또는의 대상이 되는 것들은 소괄호로 묶어야지

In [15]:
wordlist = [w for w in wsj if re.search('(ed|ing)$', w)]
print(wordlist[:10])

['62%-owned', 'Absorbed', 'According', 'Adopting', 'Advanced', 'Advancing', 'Alfred', 'Allied', 'Annualized', 'Anything']


#### 또는의 대상이 되는 string의 길이가 1일 수 있다.

In [7]:
wordlist = [w for w in wsj if re.search('(e|i)$', w)]
print(wordlist[:10])

["'re", "'ve", '100-megabyte', '100-share', '120-a-share', '18-a-share', '240-page', '30-minute', '30-share', '300-a-share']


#### nltk.FreqDist(vs for word in wsj) 이걸 먼저 생각해라
#### :wsj에 있는 단어들을 word에 넣는데, 이때 각 요소가 몇 번 나오는지를 짝 지어 줘라
#### FreqDist의 결과물은 list가 아니다

In [12]:
fdist = nltk.FreqDist(word for word in wsj)
print(fdist)
print(type(fdist))
print(len(fdist))
print(len(wsj))

<FreqDist with 12408 samples and 12408 outcomes>
<class 'nltk.probability.FreqDist'>
12408
12408


#### re.search: 뭐가 있느냐 없느냐를 찾아줌
#### re.findall: 2째input에 1째input에 해당하는 것이 있으면 그냥 다 찾아서 리스트로 만들어서 줘라
#### re.search와 re.findall이 주는 output의 type은 다르다

In [22]:
print(re.findall('ab','ababdddabkabddd'))
print(type(re.findall('ab','ababdddabkabddd')))

['ab', 'ab', 'ab', 'ab']
<class 'list'>


#### re.Match에는 len 함수를 적용할 수 없다.

In [24]:
print(re.search('ab','ababdddabkabddd'))
print(type(re.search('ab','ababdddabkabddd')))
print(len(re.search('ab','ababdddabkabddd')))

<re.Match object; span=(0, 2), match='ab'>
<class 're.Match'>


TypeError: object of type 're.Match' has no len()

#### 왜 하는가? - 이 사전에서 모음을 다 뽑아와라
#### 하나 이상이 되는 것을 뽑아오라고 할 수도 있다

In [8]:
atleastone1 = [vs for word in wsj for vs in re.findall(r'[aeiou]{1,}',word)]
print(type(atleastone1))

<class 'list'>


In [9]:
atleastone2 = [vs for word in wsj for vs in re.findall(r'[aeiou]+',word)]

In [10]:
atleastone1 == atleastone2

True

In [11]:
print(len(atleastone1))

26119


In [12]:
vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)

SyntaxError: invalid syntax (<ipython-input-12-a241399327fd>, line 1)

In [71]:
practice = [vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)]  # type: list
print(len(practice))
# print(practice)

3405


#### wsj의 첫째 요소를 word에 넣는다
#### word에서 r'[aeiou]{2,}'을 만족시키는 부분을 모두 찾아 만든 list의 요소를 아래 vs에 하나 넣는다.
#### 아래 for loop을 통해 vs에 add up해서 다 넣고나면 vs가 하나의 list로 완성되고
#### 그 완성된 list가 위vs(최종 list)에 추가되고, 다음 순서(wsj의 둘째 요소)의 for loop을 돈다.

#### vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)이 대괄호로 싸여 있어야 하는데 생략됨.
#### FrqDist의 input의 type: list
#### 아무생각 없이 모여있는 정보(3405개의 정보, 43가지 종류)에서 가장 많이 언급된 순서대로 정보들을 나열한 list를 보여줘: most_common
#### most_common에 길이를 지정하면 list, 지정하지 않으면(그냥 most_common을 쓰면) method

In [5]:
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word))
print(fdist)
print(fdist.most_common(10))
print(fdist.most_common)

<FreqDist with 43 samples and 3405 outcomes>
[('io', 549), ('ea', 476), ('ie', 331), ('ou', 329), ('ai', 261), ('ia', 253), ('ee', 217), ('oo', 174), ('ua', 109), ('au', 106)]
<bound method Counter.most_common of FreqDist({'io': 549, 'ea': 476, 'ie': 331, 'ou': 329, 'ai': 261, 'ia': 253, 'ee': 217, 'oo': 174, 'ua': 109, 'au': 106, ...})>


In [13]:
print(type(fdist.most_common(10)))
print(type(fdist.most_common))

<class 'list'>
<class 'method'>


In [3]:
word = 'supercalifragilisticexpialidocious'

#### 모음 찾아라
#### 모음 중 가장 많이 출현하는 것은? -해봐!

In [23]:
result = re.findall('[aeiou]', word)
ranking = nltk.FreqDist(result)
ranking.most_common

<bound method Counter.most_common of FreqDist({'i': 7, 'a': 3, 'u': 2, 'e': 2, 'o': 2})>

#### re.search의 입력: string
#### re.findall의 입력: string
#### re.findall: find된 대상을 찾아온다
#### 세상에서 가장 긴 이 단어에서 모음이 찾아지면 가져와라(여러개면 여러개를 다 찾아와라)

In [6]:
result = re.findall('[aeiou]', word)
print(result)

['u', 'e', 'a', 'i', 'a', 'i', 'i', 'i', 'e', 'i', 'a', 'i', 'o', 'i', 'o', 'u']


#### 모음과 모음 사이에 있는 괄호 안에 있는 character 2개를 가져와라(괄호 밖의 내용은 가져오지 않음)

In [7]:
result = re.findall('[aeiou](..)[aeiou]', word)
print(result)

['rc', 'fr', 'st', 'xp', 'ci']


#### 모음이 아닌 것과 모음이 아닌 것 사이에 character가 하나 이상 오면, 모음 아닌 것~사이에 있는 character~모음아닌 것을 가져와라 (예: s~s)
#### 소괄호가 없으므로 다 합쳐져서 나온다
#### greedy search: 가능하면 크게 범위를 가져서 가져온다(내부적으로 포함된 것을 무시한다)

In [8]:
result = re.findall('[^aeiou].+[^aeiou]', word) # greedy search
print(result)

['supercalifragilisticexpialidocious']


#### ?의 의미: optional(Zero or one of the previous item) -> 여기서는 ?가 +에 적용
#### -> 모음 아닌 것과 모음 아닌 것 사이에 일단 하나의 character가 있고, 2번째 character부터는 있어도 되고 없어도 된다
#### 바로 앞의 것과의 차이: 가능하면 좁은 범위의 것들만 가져와라.(?이 부여하는 의미) : reluctant search

In [9]:
result = re.findall('[^aeiou].+?[^aeiou]', word) # reluctant search
print(result)

['sup', 'rcal', 'frag', 'lis', 'tic', 'xpial', 'doc']


#### 모음아닌 것과 모음 아닌 것 사이에 하나의 character가 오거나 아무것도 오지 않는 경우를 가져와라

In [29]:
result = re.findall('[^aeiou].?[^aeiou]', word) # reluctant search
print(result)

['sup', 'rc', 'lif', 'rag', 'lis', 'tic', 'xp', 'lid']


#### wsj: wall street journal
#### nltk 안에 corpus 안에 treebank 안에 있는 사전을 가져옴
#### set을 통해 집합처럼 만들어서 sorted를 통해 사전화한다(분류하고, 겹치는 거 하나만 나오게 하고.)
#### wsj에 중복되는 단어도 많이 나오는데 집합으로 만들어서 a,b,c,d... 같은 sorting을 한다  : 사전화(사전에는 겹치는 단어가 없다)
#### wsj를 보면 대문자로 시작하는 단어들이(A...~Z...) 먼저 나오고 그 다음엔 소문자로 시작하는 단어들(a...~z...)이 나온다
#### sorted -> type: list

In [31]:
wsj = sorted(set(nltk.corpus.treebank.words()))
# print(wsj)
print(type(wsj))

<class 'list'>


#### 예: yielding이 들어갔다.
#### re.findall에 리스트가 만들어진다
#### re.findall의 item들이 아래의 vs에 쌓인다(2개 이상 찾으면 list가 되겠지)
#### 아래의 for loop을 다 돌아서 아래의 vs에 다 쌓이고 나면 그것이 통째로 위의 vs에 쌓인다 
#### 전체기능: 내가 원하는 조건을 만족시키는 단어들만 모아서 다시 리스트를 만든다
#### 각 item의 길이는 2보다 길 수도 있다

In [35]:
wordlist = [vs for word in wsj for vs in re.findall('[aeiou]{2,}', word)]
# print(wordlist)
print(type(wordlist))

<class 'list'>


#### FreqDist 안에 리스트를 위한 대괄호 표시를 하지 않아도 괜찮음

#### 중복되는 item이 있을 때 어떤 item의 빈도 수가 가장 많이 나오는가(FreqDist)
#### 시험 예시: Q) fdist에 사전에 해당하는 리스트를 넣었다. fdist.most_common의 output에서 오른쪽 빈도 수가 몇일까? -> A) 빈도 수가 전부 1로 나타난다.

In [38]:
fdist = nltk.FreqDist(wordlist)
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

In [39]:
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall('[aeiou]{2,}', word))
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

In [41]:
fdist1 = nltk.FreqDist(wordlist)
rank10_1=fdist1.most_common(10)

fdist2 = nltk.FreqDist(vs for word in wsj for vs in re.findall('[aeiou]{2,}', word))
rank10_2=fdist2.most_common(10)

rank10_1 == rank10_2

True

In [42]:
english_udhr = nltk.corpus.udhr.words('English-Latin1')
print(english_udhr)

['Universal', 'Declaration', 'of', 'Human', 'Rights', ...]


In [58]:
a = ['a','a','a','a']
print(' '.join(a))
print([' '.join(a)])
a = [['a','a','a','a','a'],['b','b','b']]
print([' '.join(w) for w in a])

a a a a
['a a a a']
['a a a a a', 'b b b']


#### 직접 함수를 만들자
#### def~return: 함수
#### 함수: 입력과 출력이 있어야지(입력과 출력은 모두 optional하다)
#### 함수의 이름을 내가 compress라 지정함
#### def 함수이름(입력(입력이름도 내 맘대로 정함)): .... cf. 입력이 여러 개라면 콤마로 구분
#### 함수의 내용을 쓸 때는 indent 필요(tab(space4개))
#### 여기서 word는 string이겠지(re.findall이라는 함수를 쓰고 있으니까) ('word'도 내가 만든 함수 이름)
#### return 출력
#### regexp 변수를 보면... |의 의미: or
#### ^이 대괄호 밖에 있으면 처음, ^이 대괄호 안에 있으면 not이라는 의미
#### '앞에 있는 하나 이상의 모음' or '끝에 있는 하나 이상의 모음' or '어디에 있든 모음이 아닌 것'라는 조건을 regexp에 넣어라(re.findall의 input으로 쓰임)
#### 함수의 내용: word에서 regexp에 해당하는 부분을 모두 찾아 만든 리스트를 pieces에 넣어라
#### 함수의 출력: pieces 변수 안에 있는 리스트를 이어붙여라: ''.join(pieces)
#### join의 입력: list, join의 출력: string
#### 결론: 중간 모음들을 다 죽이는 과정
#### english_udhr[:75] 안에 있는 item을 하나씩 compress 함수에 넣고, compress의 output을 list 형태로 만든다

In [44]:
regexp = '^[AEIOUaeiou]+|[AEIOUaeiou]+$|[^AEIOUaeiou]'
def compress(word):
    pieces = re.findall(regexp, word)
    return ''.join(pieces)
[compress(w) for w in english_udhr[:75]]

['Unvrsl',
 'Dclrtn',
 'of',
 'Hmn',
 'Rghts',
 'Prmble',
 'Whrs',
 'rcgntn',
 'of',
 'the',
 'inhrnt',
 'dgnty',
 'and',
 'of',
 'the',
 'eql',
 'and',
 'inlnble',
 'rghts',
 'of',
 'all',
 'mmbrs',
 'of',
 'the',
 'hmn',
 'fmly',
 'is',
 'the',
 'fndtn',
 'of',
 'frdm',
 ',',
 'jstce',
 'and',
 'pce',
 'in',
 'the',
 'wrld',
 ',',
 'Whrs',
 'dsrgrd',
 'and',
 'cntmpt',
 'fr',
 'hmn',
 'rghts',
 'hve',
 'rsltd',
 'in',
 'brbrs',
 'acts',
 'whch',
 'hve',
 'outrgd',
 'the',
 'cnscnce',
 'of',
 'mnknd',
 ',',
 'and',
 'the',
 'advnt',
 'of',
 'a',
 'wrld',
 'in',
 'whch',
 'hmn',
 'bngs',
 'shll',
 'enjy',
 'frdm',
 'of',
 'spch',
 'and']

In [46]:
regexp = '^[AEIOUaeiou]+|[AEIOUaeiou]+$|[^AEIOUaeiou]'
def compress(word):
    pieces = re.findall(regexp, word)
    return pieces
[compress(w) for w in english_udhr[:75]]

[['U', 'n', 'v', 'r', 's', 'l'],
 ['D', 'c', 'l', 'r', 't', 'n'],
 ['o', 'f'],
 ['H', 'm', 'n'],
 ['R', 'g', 'h', 't', 's'],
 ['P', 'r', 'm', 'b', 'l', 'e'],
 ['W', 'h', 'r', 's'],
 ['r', 'c', 'g', 'n', 't', 'n'],
 ['o', 'f'],
 ['t', 'h', 'e'],
 ['i', 'n', 'h', 'r', 'n', 't'],
 ['d', 'g', 'n', 't', 'y'],
 ['a', 'n', 'd'],
 ['o', 'f'],
 ['t', 'h', 'e'],
 ['e', 'q', 'l'],
 ['a', 'n', 'd'],
 ['i', 'n', 'l', 'n', 'b', 'l', 'e'],
 ['r', 'g', 'h', 't', 's'],
 ['o', 'f'],
 ['a', 'l', 'l'],
 ['m', 'm', 'b', 'r', 's'],
 ['o', 'f'],
 ['t', 'h', 'e'],
 ['h', 'm', 'n'],
 ['f', 'm', 'l', 'y'],
 ['i', 's'],
 ['t', 'h', 'e'],
 ['f', 'n', 'd', 't', 'n'],
 ['o', 'f'],
 ['f', 'r', 'd', 'm'],
 [','],
 ['j', 's', 't', 'c', 'e'],
 ['a', 'n', 'd'],
 ['p', 'c', 'e'],
 ['i', 'n'],
 ['t', 'h', 'e'],
 ['w', 'r', 'l', 'd'],
 [','],
 ['W', 'h', 'r', 's'],
 ['d', 's', 'r', 'g', 'r', 'd'],
 ['a', 'n', 'd'],
 ['c', 'n', 't', 'm', 'p', 't'],
 ['f', 'r'],
 ['h', 'm', 'n'],
 ['r', 'g', 'h', 't', 's'],
 ['h', 'v', 'e'],

In [64]:
rotokas_words = nltk.corpus.toolbox.words('rotokas.dic')
print(rotokas_words)

['kaa', 'kaa', 'kaa', 'kaakaaro', 'kaakaaviko', 'kaakaavo', 'kaakaoko', 'kaakasi', 'kaakau', 'kaakauko', 'kaakito', 'kaakuupato', 'kaaova', 'kaapa', 'kaapea', 'kaapie', 'kaapie', 'kaapiepato', 'kaapisi', 'kaapisivira', 'kaapo', 'kaapopato', 'kaara', 'kaare', 'kaareko', 'kaarekopie', 'kaareto', 'Kaareva', 'kaava', 'kaavaaua', 'kaaveaka', 'kaaveakapie', 'kaaveakapievira', 'kaaveakavira', 'kae', 'kae', 'kaekae', 'kaekae', 'kaekaearo', 'kaekaeo', 'kaekaesoto', 'kaekaevira', 'kaekeru', 'kaepaa', 'kaepie', 'kaepie', 'kaepievira', 'kaereasi', 'kaereasivira', 'kaetu', 'kaetupie', 'kaetuvira', 'kaeviro', 'kagave', 'kaie', 'kaiea', 'kaikaio', 'Kaio', 'kaipori', 'kaiporipie', 'kaiporivira', 'kairi', 'kairiro', 'kairo', 'kaita', 'kaitutu', 'kaitutupie', 'kaitutuvira', 'kakae', 'kakae', 'kakae', 'kakaevira', 'kakapikoa', 'kakapikoto', 'kakapu', 'kakapua', 'kakara', 'Kakarapaia', 'kakarau', 'Kakarera', 'kakata', 'kakate', 'kakatuara', 'kakau', 'kakauoa', 'kakavea', 'kakavoro', 'kakavu', 'kakeoto', '

#### rotokas_words의 item 1개를 w에 넣는다
#### re.findall을 통해 w에서 '[ptksvr][aeiou]' 조건에 맞는 부분을 모두 찾은 리스트가 만들어진다. 이 리스트를 '리스트2'라 하자
#### 리스트2의 첫 번째 item을 cv에 넣는다. 리스트의 첫 번째 item이 큰 리스트에 쌓인다
#### 리스트2의 두 번째 item을 cv에 넣는다. 리스트의 두 번째 item이 큰 리스트에 쌓인다
#### ...
#### 리스트의 마지막 item을 cv에 넣는다.  리스트의 마지막 item이 큰 리스트에 쌓인다

In [70]:
cvs = [cv for w in rotokas_words for cv in re.findall('[ptksvr][aeiou]', w)]
# print(cvs)
nltk.FreqDist(cvs).most_common(10)

[('ko', 420),
 ('ka', 418),
 ('ra', 187),
 ('ku', 173),
 ('to', 148),
 ('ke', 148),
 ('vi', 105),
 ('pi', 105),
 ('si', 100),
 ('ki', 94)]

#### ConditionalFreqDist의 input은 반드시 각 item의 길이가 2인 리스트여야 한다
#### 행: item의 첫 character
#### 렬: item의 둘째 character

In [63]:
cfd = nltk.ConditionalFreqDist(cvs) # CFD works pairwise only
cfd.tabulate()

    a   e   i   o   u 
k 418 148  94 420 173 
p  83  31 105  34  51 
r 187  63  84  89  79 
s   0   0 100   2   1 
t  47   8   0 148  37 
v  93  27 105  48  49 
